tests/test_device_tracker.py

```diff
@@ -0,0 +1,204 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Unit tests for the DeviceTracker and constructor
+"""
+import pytest
+
+import pennylane as qml
+from pennylane import Tracker
+
+
+class TestTrackerCoreBehaviour:
+    def test_default_initialization(self):
+        """Tests default initializalition"""
+
+        tracker = Tracker()
+
+        assert tracker.persistent == False
+        assert tracker.callback is None
+
+        assert tracker.history == dict()
+        assert tracker.totals == dict()
+        assert tracker.latest == dict()
+
+        assert tracker.tracking == False
+
+    def test_device_assignment(self):
+        """Assert gets assigned to device"""
+        dev = qml.device("default.qubit", wires=2)
+
+        tracker = Tracker(dev=dev)
+
+        assert id(dev.tracker) == id(tracker)
+
+    def test_incompatible_device_assignment(self):
+        """Assert exception raised when `supports_tracker` not True"""
+
+        class TempDevice:
+            short_name = "temp"
+
+            def capabilities(self):
+                return dict()
+
+        temp = TempDevice()
+
+        with pytest.raises(Exception, match=r"Device 'temp' does not support device tracking"):
+            Tracker(dev=temp)
+
+    def test_reset(self):
+        """Assert reset empties totals and history"""
+
+        tracker = Tracker()
+
+        tracker.totals = {"a": 1}
+        tracker.history = {"a": [1]}
+        tracker.latest = {"a": 1}
+
+        tracker.reset()
+
+        assert tracker.totals == dict()
+        assert tracker.history == dict()
+        assert tracker.latest == dict()
+
+    def test_enter_and_exit(self):
+        """Assert entering and exit work as expected"""
+
+        tracker = Tracker()
+        tracker.totals = {"a": 1}
+        tracker.history = {"a": [1]}
+        tracker.latest = {"a": 1}
+
+        returned = tracker.__enter__()
+
+        assert id(tracker) == id(returned)
+        assert tracker.tracking == True
+
+        assert tracker.totals == dict()
+        assert tracker.history == dict()
+        assert tracker.latest == dict()
+
+        tracker.__exit__(1, 1, 1)
+
+        assert tracker.tracking == False
+
+    def test_context(self):
+        """Assert works with runtime context"""
+
+        with Tracker() as tracker:
+            assert isinstance(tracker, Tracker)
+            assert tracker.tracking == True
+
+        assert tracker.tracking == False
+
+    def test_update(self):
+        """Checks update stores to history and totals"""
+
+        tracker = Tracker()
+
+        tracker.update(a=1, b="b", c=None)
+        tracker.update(a=2, b="b2", c=1)
+
+        assert tracker.history == {"a": [1, 2], "b": ["b", "b2"], "c": [None, 1]}
+
+        assert tracker.totals == {"a": 3, "c": 1}
+
+        assert tracker.latest == {"a": 2, "b": "b2", "c": 1}
+
+    def test_record_callback(self, mocker):
```

Comment:
Missing docstring

Reply:
```suggestion
    def test_record_callback_is_not_none(self, mocker):
```

Reply:
Worth adding a test for `callback is None` too.

Reply:
I'm confused as to how to test `callback is None`.  How do I test the thing I don't create isn't called?

Reply:
it could be checked that there are no new entries in the standard output/logs